programs/keeper/keeper_main.cpp

```diff
@@ -0,0 +1,449 @@
+#include <csignal>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <new>
+#include <iostream>
+#include <vector>
+#include <tuple>
+#include <string_view>
+#include <utility> /// pair
+
+#include <fmt/format.h>
+
+#include "config.h"
+#include "config_tools.h"
+
+#include <Common/StringUtils.h>
+#include <Common/getHashOfLoadedBinary.h>
+#include <Common/IO.h>
+
+#include <base/phdr_cache.h>
+#include <base/coverage.h>
+
+
+int mainEntryClickHouseKeeper(int argc, char ** argv);
+#if ENABLE_CLICKHOUSE_KEEPER_CONVERTER
+int mainEntryClickHouseKeeperConverter(int argc, char ** argv);
+#endif
+#if ENABLE_CLICKHOUSE_KEEPER_CLIENT
+int mainEntryClickHouseKeeperClient(int argc, char ** argv);
+#endif
+
+namespace
+{
+
+using MainFunc = int (*)(int, char**);
+
+/// Add an item here to register new application
+std::pair<std::string_view, MainFunc> clickhouse_applications[] =
+{
+    // keeper
+    {"keeper", mainEntryClickHouseKeeper},
+#if ENABLE_CLICKHOUSE_KEEPER_CONVERTER
+    {"converter", mainEntryClickHouseKeeperConverter},
+    {"keeper-converter", mainEntryClickHouseKeeperConverter},
+#endif
+#if ENABLE_CLICKHOUSE_KEEPER_CLIENT
+    {"client", mainEntryClickHouseKeeperClient},
+    {"keeper-client", mainEntryClickHouseKeeperClient},
+#endif
+
+};
+
+int printHelp(int, char **)
+{
+    std::cerr << "Use one of the following commands:" << std::endl;
+    for (auto & application : clickhouse_applications)
+        std::cerr << "clickhouse " << application.first << " [args] " << std::endl;
+    return -1;
+}
+
+
+enum class InstructionFail : uint8_t
```

Comment:
Is the SIMD detection code l. 63 - 222 relevant for Keeper? If yes, we can perhaps move it to a common header (as it duplicates code in programs/main.cpp), or add a comment in both files with a note to keep the code in-sync.

EDIT: Reading on, the same applies to more code below.

Reply:
> Is the SIMD detection code l. 63 - 222 relevant for Keeper?

Not sure, probably not but I don't see an issue for having it included just in case some of our lib uses it.
I can extract the common parts ofc